### PR TITLE
Clickhouse driver [WIP]

### PR DESCRIPTION
**Pull request by ei-grad**
Work in progress on ClickHouse driver.

It works, but see https://github.com/metabase/metabase/pull/8722#issuecomment-448017028.

Uberjar to test: https://github.com/ei-grad/metabase/releases/download/v0.31.2-ClickHouse1/metabase.jar

Related: #3332 #8491

### CODE REVIEW

File: src/metabase/driver/clickhouse.clj [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+                             :expressions
+                             :expression-aggregations})})
+  sql/ISQLDriver
+  (merge (sql/ISQLDriverDefaultsMixin)
+         {:active-tables             sql/post-filtered-active-tables
+          :column->base-type         (u/drop-first-arg column->base-type)
+          :connection-details->spec  (u/drop-first-arg connection-details->spec)
+          :date                      (u/drop-first-arg date)
+          :excluded-schemas          (constantly #{"system"})
+          :quote-style               (constantly :postgres)
```
[the remark this answers is not in the document]
Comment (ei-grad) [reply src/metabase/driver/clickhouse.clj]: I think this line should be just deleted, there is no ":postgres" quoting in honeysql, iirc. The PostgreSQL driver uses the default `:ansi`, and it also should be ok for clickhouse. Anyway, thanks for pointing this out, I have already fixed it but then had troubles with merge and lost this change during the rebase/squash. And why do you suggesting the mysql quoting? What's the difference?

[the remark this answers is not in the document]
Comment (ei-grad) [reply src/metabase/driver/clickhouse.clj]: Oh, really, it looks that the backtick thing is MySQL-specific. Don't know why I did remember wrong.

**Comment by enqueue** [comment]
Regarding the NPE: This is probably due to the JDBC driver returning a `null` result set for queries starting with comments, e. g. `-- MetaBase\n SELECT 42` I created PR yandex/clickhouse-jdbc#256 to counter this.

Regarding the test data, I would like to ask @camsaul for advice: The `test-data/checkins` table has a `date` column defined with type `base-type :type/Date`. Then data is loaded via a timestamp, e.g. `#inst "2014-09-15T07` resulting in `2014-09-15 07:00:00` on the wire. This is rejected by the ClickHouse server because it cannot parse this input as a Date ("Date must be in YYYY-MM-DD format"). What is expected of the driver in this case? What would be the right place in the code to adjust the format of the timestamp depending on the target column type when loading data?

**Comment by enqueue** [comment]
@ei-grad is there a way for me to create a PR against this branch that is already rebased on master? See changes in [enqueue/clickhouse_update](https://github.com/enqueue/metabase/tree/clickhouse_update)

We are far away from completion (ca. 40 tests still failing), but if you compile JDBC driver including your changes yandex/clickhouse-jdbc#247 you will be able to play around with the database and perhaps improve the driver further ;-)

**Comment by ei-grad** [comment]
Rebased, moved the `project.clj` change to separate commit to easier rebase in future.

**Comment by ei-grad** [comment]
@enqueue rebased, moved the project.clj change to separate commit to easier rebase in future.

**Comment by ei-grad** [comment]
Merged, removed the `TODO_clickhouse.txt` file, not sure if it should go in the source tree, it is better to put it here, imo.

**Comment by ei-grad** [comment]
It is a bit strange, but when I run the tests with `ENGINES=h2,clickhouse lein test` the result is:

Ran 3446 tests containing 3342 assertions in 91516 msecs
1 failures, 8 errors.

Did something changed with running the tests?..

**Comment by ei-grad** [comment]
There are some buggy empty extra fields in clickhouse driver configuration UI:
screenshot_2018-11-26 metabase

**Comment by ei-grad** [comment]
@enqueue I didn't have a look to the latest driver structure refactoring yet, do you have any ideas why this extra fields appeared? ^

File: src/metabase/driver/clickhouse.clj [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+            [metabase.mbql.util :as mbql.u]
+            [metabase.models.field :as field]
+            [metabase.util :as u]
+            [metabase.util
+             [honeysql-extensions :as hx]
+             [ssh :as ssh]]))
+
+(driver/register! :clickhouse, :parent :sql-jdbc)
+
+(def ^:private default-base-types
```
Comment (Bugagazavr) [inline src/metabase/driver/clickhouse.clj]: clickhouse also supports `Decimal` https://clickhouse.yandex/docs/ru/data_types/decimal/

**Comment by enqueue** [comment]
@ei-grad thanks for your help! Let's try to get this PR ready for christmas...

* Will remove  `TODO_clickhouse.txt` when we are done with this branch (unless we forget ;-) )
* You have to run the tests using `DRIVERS=h2,ClickHouse` instead of `ENGINE`
* I do not know about the buggy extra fields. Are these specific to ClickHouse driver?

File: src/metabase/driver/clickhouse.clj (hunk 2 above)
Comment (enqueue) [reply src/metabase/driver/clickhouse.clj]: Thank you! Added the data type to the test.

**Comment by enqueue** [comment]
Hi @camsaul,

we spent some time trying to get the ClickHouse driver to work, based on the great foundations laid by you and many others. Some issues could be resolved by enhancements to the JDBC driver. We are down to 3 failures and 4 errors when running the CI tests, but I have reached my limits somewhat and would like to ask for your help.

# Aliases versus field names

When the field name is used as alias, ClickHouse gets a little bit confused when the field is used again at a later point in the query.

Example:

```
SELECT 
    `sad-toucan-incidents`.incidents.id AS id, 
    `sad-toucan-incidents`.incidents.severity AS severity, 
    toDateTime(`sad-toucan-incidents`.incidents.timestamp / 1000) AS timestamp
FROM `sad-toucan-incidents`.incidents 
WHERE toDateTime(`sad-toucan-incidents`.incidents.timestamp / 1000) = '2015-06-02 00:00:00'
ORDER BY timestamp ASC
LIMIT 10
```
results in `Illegal types DateTime and UInt16 of arguments of function divide.` In the database, `timestamp` is modelled as a 64 bit integer, so this is not the issue. The query works if a different alias than `timestamp` is used. I tried overriding `field->alias` but this created many more errors.

Test | Line
---- | ----
metabase.query-processor-test.date-bucketing-test |874
metabase.query-processor-test.date-bucketing-test |878
metabase.query-processor-test.unix-timestamp-test |12

# Date ranges

A current effective query looks like this:

```SELECT COUNT(*) FROM `test-data`.`checkins` WHERE toDate(`test-data`.`checkins`.`date`) BETWEEN '2018-12-01 00:00:00' AND '2018-12-31 00:00:00'```

The server does not like this, saying: `String is too long for Date: 2018-12-01 00:00:00`.

Why is the date cast to Timestamp in `metabase.query-processor.middleware.parameters.sql`? A Date would probably suffice here. Alternatively, an invocation of `(sql.qp/date driver/*driver* :day xxxx)` might be necessary. I read in #4043 that this would have a huge performace impact, though. We also do not want to override `sql/->prepared-substitution` _generally_ for Timestamps.

Test | Line |
---- | ----
 metabase.query-processor.middleware.parameters.sql-test |572
 metabase.query-processor.middleware.parameters.sql-test |601
 metabase.query-processor.middleware.parameters.sql-test |615

# Something else?

Test | Line |
---- | ---- 
metabase.api.table-test| 703

# Time

ClickHouse does not know about a Time data type. I commented out the tests, but perhaps you have a good hint on how to get this working properly, despite the lack of a dedicated data type.

# Modular drivers

We saw that some drivers are now distributed as modules and we'll need to rebase and do some service loader magic. Do you see any obvious obstacles for this?

Thank you very much in advance for any hints!

**Comment by CheatEx** [comment]
Hi. Does anyone working on finishing this effort? Specifically on porting to new plugin system?

**Comment by enqueue** [comment]
Hello @CheatEx I have been working on this driver, and refactored it for the new module structure. Unfortunately I cannot update this PR myself (see ei-grad/metabase#6). If you want to check out a more recent version, see [clickhouse_module_2](https://github.com/enqueue/metabase/tree/clickhouse_module_2) branch. What should I do?

**Comment by zeronewb** [comment]
imho it should be rebased to new PR :/
@ei-grad, @enqueue  WDYT?

**Pull request closed without merging**